chore: lock free info command with replicaof v2

kostasrim opened the pull request:
Following #5774, this PR removes locking the `replica_of_mu_` from info command and uses the thread locals && replica of v2 algorithm.


No more locking and blocking for INFO command during `replicaof` or `takeover` 🥳 🎉 🌮

kostasrim committed:
chore: lock free info command with replicaof v2

Signed-off-by: kostas <kostas@dragonflydb.io>

kostasrim committed:
cleanup

kostasrim committed:
cleanup

kostasrim committed:
tests

kostasrim committed:
test old path

kostasrim committed:
fix

kostasrim commented:
With all the changes around replication, I will follow up with a tidy PR 😄

romange commented on src/server/protocol_client.cc:
```diff
@@ -119,13 +120,6 @@ ProtocolClient::ProtocolClient(ServerContext context) : server_context_(std::mov
 ProtocolClient::~ProtocolClient() {
   exec_st_.JoinErrorHandler();
 
-  // FIXME: We should close the socket explictly outside of the destructor. This currently
```
I am afraid the crash is a symptom of a bug in another place. Can you please create a branch that crashes and give me the instructions on how to cause the crash? I would like to inspect it myself

romange replied:
`What is more, the core dump shows that tl_replica and its copy, have a different ref counted object` - what is the copy of tl_replica?

kostasrim replied:
And I did not have the time to verify this because I literally left for holidays the evening I came up with this hypothesis 😄

romange commented on src/server/server_family.cc:
```diff
@@ -3470,6 +3505,9 @@ void ServerFamily::AddReplicaOf(CmdArgList args, const CommandContext& cmd_cntx)
   add_replica->StartMainReplicationFiber(nullopt);
   cluster_replicas_.push_back(std::move(add_replica));
+  service_.proactor_pool().AwaitFiberOnAll(
+      [this](auto index, auto* cntx)
```
this is not the right way to do it. `cluster_replicas_` can change in between. 
instead you should capture it in the callback by value instead of `this`.

romange commented on src/server/server_family.cc:
```diff
@@ -3126,16 +3144,24 @@ string ServerFamily::FormatInfoMetrics(const Metrics& m, std::string_view sectio
     // Please note that we we do not use Metrics object here.
-    {
+    if (use_replica_of_v2_) {
+      // Deep copy because tl_replica might be overwritten inbetween
```
I find confusing the "Deep copy" term here. 
1. You copy shared_ptr which is not deep copy, imho.
2. But then you `cluster_replicas` which is indeed a (deep) copy of a vector.

romange reviewed (commented):
1. Please prepare a "crash" branch for me to investigate
2. If outgoing_slot_migration change is a bug, I would like to ask to send it in another PR. Also, retire `info_replication_valkey_compatible` flag in the same PR by removing the "false" branch. Lets submit changes we are more confident about first.

romange commented on src/server/protocol_client.h:
```diff
@@ -107,7 +107,8 @@ class ProtocolClient {
   auto* Proactor() const {
-    return sock_->proactor();
+    DCHECK(socket_thread_);
+    return socket_thread_;
```
Ok, I think I understand and I think it's  a problem in the design and you totally ignore it.

`ProtocolClient/Replica` are not thread safe objects.
by caching the object pointers in all the threads, you access them freely but the access is not thread safe. This probably creates a bug that causes crashes.

kostasrim commented:
> There is a problem with this design. You use shared_ptr to be able to access the object even if it's "released" in another thread. But `ProtocolClient` is _designed_ to be destructed in its own thread. so suppose, `ProtocolClient` is created in thread A, and thread B holds the pointer to that object, and then A destroys the reference to ProtocolClient, the thread B will hold the only reference and essentially will call the destructor of ProtocolClient in the wrong thread.

I don't see any issue in destructing in a separate thread because by the time we call the destructor of `~ProtocolClient` we have cleaned our resources. Specifically:

1. shard_flows_ is empty
2. sync_fb_ and acks_fb_ should have joined (I can add DCHECKS for this)
3. base class ProtocolClient::sock_ is already `Shutdown() && Closed()` properly on the right thread so there is no IO involved or anything else

The destructor is "trivial" from that aspect (of course we had preemption because of the Await which I removed)
